service/history/workflow/update/registry_test.go

```diff
@@ -43,8 +45,34 @@ type (
 		*require.Assertions
 		suite.Suite
 	}
+
+	mockStore struct {
+		Storage
+		AddWorkflowExecutionUpdateAcceptedEventFunc  func(string, *updatepb.Acceptance) (*historypb.HistoryEvent, error)
+		AddWorkflowExecutionUpdateCompletedEventFunc func(*updatepb.Response) (*historypb.HistoryEvent, error)
+		GetAcceptedWorkflowExecutionUpdateIDsFunc    func(context.Context) ([]string, error)
```

Comment:
optional: I think you should just provide a fake object which records the relevant state instead of using lambdas. That's also helpful for other packages that need the fake.

Reply:
Have you seen this style of mocking? Avoiding all the recording stuff is exactly the point, otherwise I'd generate a gomock (though I'm not a fan of gomock or providing public mocks). There's some exposition [here](https://medium.com/@matryer/meet-moq-easily-mock-interfaces-in-go-476444187d10).

Reply:
I have seen it, and I definitely prefer the Moq style over gomock, but I prefer the embedded interface approach over them all because it requires no generated code. However, above all of the approaches, I prefer just not mocking any first-party code, and instead mocking the I/O that first-party code uses. With that practice, you don't need to write integration tests because the unit tests are integration tests--they just don't use any real IO.